sympy/tensor/array/array_comprehension.py

```diff
@@ -30,184 +31,224 @@ class ArrayComprehension(Basic):
     >>> b = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k))
     >>> b.doit()
     ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k))
-
     """
     def __new__(cls, function, *symbols, **assumptions):
         if any(len(l) != 3 or None for l in symbols):
             raise ValueError('ArrayComprehension requires values lower and upper bound'
                               ' for the expression')
         arglist = [sympify(function)]
         arglist.extend(cls._check_limits_validity(function, symbols))
-        obj = Basic.__new__(cls, *arglist, **assumptions)
-        obj._function = obj._args[0]
-        obj._limits = obj._args[1:]
-        obj._shape = cls._calculate_shape_from_limits(obj._limits)
-        obj._rank = len(obj._shape)
-        obj._loop_size = cls._calculate_loop_size(obj._shape)
```

Comment:
In fact I was doing so based on the assumption that this array is immutable. So I thought it would be better to store the value once it is created than re-compute it every time it is called. 
For example. you don't need to `_calculate_shape_from_limits` every time the `shape`  is being called.  Of course, some of these could surely be eliminated if it really causes a worry about memory.

Reply:
Consider this from the current master,
```
>>> from sympy.tensor.array import ArrayComprehension
>>> from sympy import symbols
>>> i, j, k = symbols('i, j, k')
>>> A = ArrayComprehension(i+j+k, (i, 1, 2), (j, 1, 3), (k, 1, 4))
>>> A.doit()
[[[3, 4, 5, 6], [4, 5, 6, 7], [5, 6, 7, 8]], [[4, 5, 6, 7], [5, 6, 7, 8], [6, 7, 8, 9]]]
>>> A.function
i + j + k
>>> A._function = i*j*k
>>> A.function
i*j*k
>>> A.doit()
[[[1, 2, 3, 4], [2, 4, 6, 8], [3, 6, 9, 12]], [[2, 4, 6, 8], [4, 8, 12, 16], [6, 12, 18, 24]]]
```
The above code allows the change of `A.function` using `A._function`. As a user, I won't expect that since `ArrayComprehension`. In short, current implementation leaves a scope for changing the attributes. This PR doesn't allow that. 
I don't know much about `sympy.tensor.array` but such loopholes shouldn't arise. 
One more point to consider is, that using extra attributes not only takes extra memory space but also, the access time is increased and we to optimize we should not use `__slots__` here as `ArrayComprehension` doesn't inherit directly from `object`. So, according to my point of view reducing the number of attributes will be good.

If you still want to use these attributes, like, `_function`, `_rank`, then I think we should use `_new` method to initalise those attributes as done `ImmutableDenseNDimArray`. Thanks for the comment.

Reply:
Well, private objects starting with `_` should not be used by end-users. I don't see this as a problem.

Reply:
> Well, private objects starting with `_` should not be used by end-users. I don't see this as a problem.

Agreed on this, `_` are considered to be private in python. What's the reason behind doing all the computations while creating the object? Why don't we do the required work when the time comes?

Reply:
We are guaranteed to do it once. When you create the object you are supposed to check if you are creating something meaningful, so just do all pre-computations.